internal/db/fetcher/indexer_iterators.go

```diff
@@ -499,100 +485,109 @@ type fieldFilterCond struct {
-	for i := range f.indexedFields {
-		fieldInd := f.mapping.FirstIndexOfName(f.indexedFields[i].Name)
+	// we process first the conditions that match composite index fields starting from the first one
+	for i := range f.indexDesc.Fields {
+		indexedField := f.indexedFields[i]
+		fieldInd := f.mapping.FirstIndexOfName(indexedField.Name)
 		found := false
-		// iterate through conditions and find the one that matches the current field
-		for filterKey, indexFilterCond := range f.indexFilter.Conditions {
-			propKey, ok := filterKey.(*mapper.PropertyIndex)
-			if !ok || fieldInd != propKey.Index {
-				continue
-			}
+		var err error
 
-			found = true
-
-			fieldDef := f.indexedFields[slices.IndexFunc(f.indexedFields, func(f client.FieldDefinition) bool {
-				return int(f.ID) == fieldInd
-			})]
-
-			condMap := indexFilterCond.(map[connor.FilterKey]any)
-
-			jsonPath := client.JSONPath{}
-			if fieldDef.Kind == client.FieldKind_NILLABLE_JSON {
-			jsonPathLoop:
-				for {
-					for key, filterVal := range condMap {
-						prop, ok := key.(*mapper.ObjectProperty)
-						if !ok {
-							// if filter contains an array condition, we need to append index 0 to the json path
-							// to limit the search only to array elements
-							op, ok := key.(*mapper.Operator)
-							if ok && isArrayCondition(op.Operation) {
-								if op.Operation == compOpNone {
-									// if the array condition is _none it doesn't make sense to use index  because
-									// values picked by the index is random guessing. For example if we have doc1
-									// with array of [3, 5, 1] and doc2 with [7, 4, 8] the index first fetches
-									// value 1 of doc1, let it go through the filter and then fetches value 3 of doc1
-									// again, skips it (because it cached doc1 id) and fetches value 4 of doc2, and
-									// so on until it exhaust all prefixes in ascending order.
-									// It might be even less effective than just scanning all documents.
-									return nil, nil
+		filter.TraverseProperties(
+			f.indexFilter.Conditions,
+			func(prop *mapper.PropertyIndex, condMap map[connor.FilterKey]any) bool {
+				if fieldInd != prop.Index {
+					return true
+				}
+
+				jsonPath := client.JSONPath{}
+				if indexedField.Kind == client.FieldKind_NILLABLE_JSON {
+				jsonPathLoop:
+					for {
+						for key, filterVal := range condMap {
+							prop, ok := key.(*mapper.ObjectProperty)
+							if !ok {
+								// if filter contains an array condition, we need to append index 0 to the json path
+								// to limit the search only to array elements
+								op, ok := key.(*mapper.Operator)
+								if ok && isArrayCondition(op.Operation) {
+									jsonPath = jsonPath.AppendIndex(0)
 								}
-								jsonPath = jsonPath.AppendIndex(0)
+								break jsonPathLoop
 							}
-							break jsonPathLoop
+							jsonPath = jsonPath.AppendProperty(prop.Name)
+							condMap = filterVal.(map[connor.FilterKey]any)
```

Comment:
todo: Why is this cast safe (no `ok` check)?  Please add your answer as a code comment.